### PR TITLE
Dynamic Parameters v2

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
### Description

DynamicParameters V2 finally! Includes new generation script and updates to all accesses of the old parameter system to access the new one. Unit testing pending, hence the WIP. 

Please make sure to review these files:
- `generate_parameters_v2.py` script
- `config_utils.hpp`
- `config_test.cpp`
-  `parameter.h`

### Testing Done
Unit tested, iterates over generated structure and makes sure the yaml matches
Manually tested w/ visualizer and Grsim, everything seems to work

### Resolved Issues
#811 

### Length Justification

The changes aren't very hard to follow its replacing `::` with `->` that takes up most of the PR. The actual gen script and `config_utils.hpp` file aren't long.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### Sample generated class in config.hpp
~~~c++
class AIConfig : public Config
{
   public:
    AIConfig()
    {
        RefboxConfig_config = std::make_shared<RefboxConfig>();
std::vector<bool>VisionFlippingFilterEnabled_param_options = std::vector<bool>{};
        VisionFlippingFilterEnabled_param = std::make_shared<Parameter<bool>>("VisionFlippingFilterEnabled", true,VisionFlippingFilterEnabled_param_options,
                                std::nullopt, std::nullopt);
std::vector<bool>RunAI_param_options = std::vector<bool>{};
        RunAI_param = std::make_shared<Parameter<bool>>("RunAI", true,RunAI_param_options,
                                std::nullopt, std::nullopt);
std::vector<bool>OverrideAIPlay_param_options = std::vector<bool>{};
        OverrideAIPlay_param = std::make_shared<Parameter<bool>>("OverrideAIPlay", false,OverrideAIPlay_param_options,
                                std::nullopt, std::nullopt);
std::vector<std::string>CurrentAIPlay_param_options = std::vector<std::string>{"Example Play","Stop Play","Halt Play","Corner Kick Play","Free Kick Play","Defense Play","KickoffEnemy Play","KickoffFriendly Play","Enemy Freekick Play","Penalty Kick Play","Penalty Kick Enemy Play","ShootOrChip Play"};
        CurrentAIPlay_param = std::make_shared<Parameter<std::string>>("CurrentAIPlay", "Example Play",CurrentAIPlay_param_options,
                                std::nullopt, std::nullopt);
        mutable_internal_param_list = {VisionFlippingFilterEnabled_param,
RunAI_param,
OverrideAIPlay_param,
CurrentAIPlay_param,
RefboxConfig_config};
        immutable_internal_param_list = {std::const_pointer_cast<const Parameter<bool>>(VisionFlippingFilterEnabled_param),
std::const_pointer_cast<const Parameter<bool>>(RunAI_param),
std::const_pointer_cast<const Parameter<bool>>(OverrideAIPlay_param),
std::const_pointer_cast<const Parameter<std::string>>(CurrentAIPlay_param),
std::const_pointer_cast<const RefboxConfig>(RefboxConfig_config)};
    }
    const std::shared_ptr<const Parameter<bool>> VisionFlippingFilterEnabled () const
           {
               return std::const_pointer_cast<const Parameter<bool>>(VisionFlippingFilterEnabled_param);
           }

           const std::shared_ptr<Parameter<bool>> mutableVisionFlippingFilterEnabled ()
           {
               return VisionFlippingFilterEnabled_param;
           }
        
const std::shared_ptr<const Parameter<bool>> RunAI () const
           {
               return std::const_pointer_cast<const Parameter<bool>>(RunAI_param);
           }

           const std::shared_ptr<Parameter<bool>> mutableRunAI ()
           {
               return RunAI_param;
           }
        
const std::shared_ptr<const Parameter<bool>> OverrideAIPlay () const
           {
               return std::const_pointer_cast<const Parameter<bool>>(OverrideAIPlay_param);
           }

           const std::shared_ptr<Parameter<bool>> mutableOverrideAIPlay ()
           {
               return OverrideAIPlay_param;
           }
        
const std::shared_ptr<const Parameter<std::string>> CurrentAIPlay () const
           {
               return std::const_pointer_cast<const Parameter<std::string>>(CurrentAIPlay_param);
           }

           const std::shared_ptr<Parameter<std::string>> mutableCurrentAIPlay ()
           {
               return CurrentAIPlay_param;
           }
        const std::shared_ptr<const RefboxConfig> getRefboxConfig () const
           {
               return std::const_pointer_cast<const RefboxConfig>(RefboxConfig_config);
           }

           const std::shared_ptr<RefboxConfig> getMutableRefboxConfig ()
           {
               return RefboxConfig_config;
           }
        

    const std::string name() const
    {
        return "AIConfig";
    }

    const MutableParameterList& getMutableParameterList()
    {
        return mutable_internal_param_list;
    }

    const ParameterList& getParameterList() const
    {
        return immutable_internal_param_list;
    }

   private:
        MutableParameterList mutable_internal_param_list;
        ParameterList immutable_internal_param_list;
        std::shared_ptr<Parameter<bool>> VisionFlippingFilterEnabled_param;
std::shared_ptr<Parameter<bool>> RunAI_param;
std::shared_ptr<Parameter<bool>> OverrideAIPlay_param;
std::shared_ptr<Parameter<std::string>> CurrentAIPlay_param;std::shared_ptr<RefboxConfig> RefboxConfig_config;
};
~~~

### CODE REVIEW

**Comment by akhilveeraghanta** [comment]
`full_system` is segfaulting rn, I __think__ its to do w/ static initialization and it seems like the `world` object is being initialized first before the `DynamicParmeters` object. Will look into that, but the fix shouldn't effect the script or the rest of the changes, so this is ready for "WIP" review

edit: once it stops segfaulting and ive got some tests in here ill remove the WIP

**Comment by garethellis0** [comment]
@jonathanlew @kelvinkoon @jared-paul - ya'll get the first pass! 

(@jared-paul - added you because I think it's valuable to get some perspective from someone who hasn't been super involved with the whole design process for this, let us know if you have any questions in #proj_dynamic_params on slack)

**Comment by jonathanlew** [comment]
> `full_system` is segfaulting rn, I **think** its to do w/ static initialization and it seems like the `world` object is being initialized first before the `DynamicParmeters` object. Will look into that, but the fix shouldn't effect the script or the rest of the changes, so this is ready for "WIP" review
> 
> edit: once it stops segfaulting and ive got some tests in here ill remove the WIP

Is this related to #967?

File: src/software/util/parameter/dynamic_parameters_test.cpp [hunk 1]
```diff
@@ -1,17 +1,17 @@
-#include <gtest/gtest.h>
+rinclude<gtest / gtest.h>
```
Comment (jonathanlew) [inline src/software/util/parameter/dynamic_parameters_test.cpp]: Typo?

File: src/software/util/parameter/parameter.h [hunk 2]
```diff
@@ -7,6 +7,7 @@
 #include <vector>
 
 /**
+ * TODO update this comment
```
Comment (jonathanlew) [inline src/software/util/parameter/parameter.h]: Todo?

**Review by jonathanlew — CHANGES_REQUESTED** [review]
So I need to build this branch to see what the auto generated files look like?

File: src/software/util/parameter/dynamic_parameters_test.cpp (hunk 1 above)
Comment (akhilveeraghanta) [reply src/software/util/parameter/dynamic_parameters_test.cpp]: this file is unused, nuked it

File: src/software/util/parameter/parameter.h (hunk 2 above)
Comment (akhilveeraghanta) [reply src/software/util/parameter/parameter.h]: Fixed

File: src/software/util/parameter/parameter.h [hunk 3]
```diff
@@ -68,6 +67,7 @@ class Parameter
     void setValue(const T new_value)
     {
         std::scoped_lock value_lock(this->value_mutex_);
+        std::cerr << this->name_ << " just got a new value: " << new_value << std::endl;
```
Comment (akhilveeraghanta) [inline src/software/util/parameter/parameter.h]: will remove this w/  first round of pr fixes

**Comment by akhilveeraghanta** [comment]
Guys don't do a full review until I remove "WIP" realized there is a bit more work to do than just tests, will tag everyone when the PR is ready.

**Comment by akhilveeraghanta** [comment]
@kelvinkoon @jonathanlew @jared-paul @MathewMacDougall @garethellis0 , sorry for the confusion its ready for review now :shipit: thanks guys

**Comment by garethellis0** [comment]
@akhilveeraghanta - I see the example you posted has mutable stuff in it, what's the process for getting the non-mutable config?

**Comment by akhilveeraghanta** [comment]
You would need to cast the highlevel config using a `const_pointer_cast` 
~~~c++
    6     extern const std::shared_ptr<ThunderbotsConfig> MutableDynamicParameters =
    7         std::make_shared<ThunderbotsConfig>();
    8     extern const std::shared_ptr<const ThunderbotsConfig> DynamicParameters =
    9         std::const_pointer_cast<const ThunderbotsConfig>(MutableDynamicParameters);
~~~

File: src/software/util/parameter/config_test.cpp [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    const std::shared_ptr<ThunderbotsConfig> MutableDynamicParameters =
+        std::make_shared<ThunderbotsConfig>();
+
+    // This creates an immutable ThunderbotsConfig w/ proper const correctnesss
+    const std::shared_ptr<const ThunderbotsConfig> DynamicParameters =
+        std::const_pointer_cast<const ThunderbotsConfig>(MutableDynamicParameters);
+
+    mutate_all_parameters(MutableDynamicParameters);
+    assert_mutation(DynamicParameters, config_yaml);
+}
```
Comment (garethellis0) [inline src/software/util/parameter/config_test.cpp:241]: This is super slick :heavy_check_mark:

File: src/software/util/parameter/config_utils.hpp [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+// parameters/configs in ParameterVariant can not be mutated.
+using ParameterVariant = std::variant<
+    std::shared_ptr<const Parameter<int>>, std::shared_ptr<const Parameter<bool>>,
+    std::shared_ptr<const Parameter<double>>,
+    std::shared_ptr<const Parameter<std::string>>, std::shared_ptr<const Config>>;
+
+// List of ParameterVariants (immutable)
+using ParameterList = std::vector<ParameterVariant>;
+
+// List of MutableParameterVaraints (mutable)
```
Comment (garethellis0) [inline src/software/util/parameter/config_utils.hpp]:
``` suggestion
// List of MutableParameterVariants (mutable)
```

File: src/software/util/parameter/generate_parameters_v2.py [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+    :param parameter_name: the name of the parameter
+    :param parameter_description: the yaml configuration for the parameter 
+    :type parameter_name: str
+    :type parameter_description: dict
+
+    """
+
+    def __init__(self, parameter_name: str, parameter_description: dict):
+
+        """INITIALIZER"""
```
Comment (garethellis0) [inline src/software/util/parameter/generate_parameters_v2.py]: I'm not sure this comment adds much value.

File: src/software/util/parameter/generate_parameters_v2.py [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+    """
+
+    # stores all the configurations in a "flat" structure
+    # globally to print all the classes and their
+    # forward declarations
+    all_configs = []
+
+    def __init__(self, config_name: str,  config_description: dict):
+
+        """INITIALIZER"""
```
Comment (garethellis0) [inline src/software/util/parameter/generate_parameters_v2.py]: Not sure this adds much value either, same goes for other instances?

File: src/software/util/parameter/generate_parameters_v2.py [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+
+
+#######################################################################
+#                                MAIN                                 #
+#######################################################################
+if __name__ == '__main__':
+
+    # create config
+    ThunderbotsConfig =\
+        Config('ThunderbotsConfig', load_configuration(sys.argv[1:-1]))
```
Comment (garethellis0) [inline src/software/util/parameter/generate_parameters_v2.py:337]: Are we just loading one config? Or what is this top-level name for?

File: src/software/util/parameter/parameter.h [hunk 9]
```diff
@@ -3,17 +3,16 @@
...
 #include <vector>
 
 /**
  * This class defines a dynamic parameter, meaning the parameter
- * value can be changed during runtime. Although the class is templated, it is only meant
- * to support the types also supported by the ROS Parameter Server.
+ * value can be changed during runtime.
  *
- * See http://wiki.ros.org/Parameter%20Server for the list of types
+ * In our codebase, we currently support bool, int32_t, double, and strings
```
Comment (garethellis0) [inline src/software/util/parameter/parameter.h]: why `int32_t` over `int`?

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Looking good, no major comments

File: src/software/util/parameter/config_utils.hpp (hunk 5 above)
Comment (akhilveeraghanta) [reply src/software/util/parameter/config_utils.hpp]: fixed

File: src/software/util/parameter/generate_parameters_v2.py (hunk 6 above)
Comment (akhilveeraghanta) [reply src/software/util/parameter/generate_parameters_v2.py]: yeah my snipped popped it here didn't notice it, its gone now

File: src/software/util/parameter/generate_parameters_v2.py (hunk 7 above)
Comment (akhilveeraghanta) [reply src/software/util/parameter/generate_parameters_v2.py]: removed

File: src/software/util/parameter/generate_parameters_v2.py (hunk 8 above)
Comment (akhilveeraghanta) [reply src/software/util/parameter/generate_parameters_v2.py:337]: all the yaml files are loaded into this one config, so you could access all the top level configs from this one config

File: src/software/util/parameter/parameter.h (hunk 9 above)
Comment (akhilveeraghanta) [reply src/software/util/parameter/parameter.h]: it was just a ROS quirk, we just use `int` now, updated the comment

**Comment by garethellis0** [comment]
Nice job addressing comments so quickly! Just the one left.

File: src/software/ai/hl/stp/evaluation/indirect_chip.cpp [hunk 10]
```diff
@@ -50,30 +52,40 @@ std::optional<Point> Evaluation::findTargetPointForIndirectChipAndChase(
...
+        std::optional<LegacyTriangle> largest_triangle =
+            getLargestValidTriangle(triangles,
+                                    Util::DynamicParameters->getEvaluationConfig()
+                                        ->getIndirectChipConfig()
+                                        ->MinChipTriArea()
+                                        ->value(),
+                                    Util::DynamicParameters->getEvaluationConfig()
+                                        ->getIndirectChipConfig()
+                                        ->MinChipTriEdgeLen()
+                                        ->value());
```
[the remark this answers is not in the document]
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/evaluation/indirect_chip.cpp]: Did we wan't to actually do this across all the files? or did you mean just for this file?

File: src/software/util/parameter/config_test.cpp (hunk 4 above)
Comment (akhilveeraghanta) [reply src/software/util/parameter/config_test.cpp:241]: lol just wanted to 🅱️e safe

File: src/WORKSPACE [hunk 11]
```diff
@@ -61,6 +61,14 @@ new_git_repository(
     remote = "https://github.com/saebyn/munkres-cpp.git",
 )
 
+# yaml cpp parser for dynamic parameters test
+git_repository(
+    name = "yaml-cpp",
+    commit = "f531f8a8c3f8ca0923f9219c72fcca511555f824",
+    remote = "https://github.com/jbeder/yaml-cpp.git",
+    shallow_since = "1571445859 -0700",
```
Comment (jonathanlew) [inline src/WORKSPACE]: can we clarify why this commit?

File: src/software/ai/hl/stp/evaluation/possession_test.cpp [hunk 12]
```diff
@@ -190,22 +190,30 @@ TEST(PossessionEvaluationTest, team_had_possession_half_second_ago)
...
+                                       ->PossessionBufferTimeSeconds()
+                                       ->value()));
     robot1.updateState(
         Point(-2, 0), Vector(), Angle::quarter(), AngularVelocity::zero(),
-        Timestamp::fromSeconds(2 + Util::DynamicParameters::Evaluation::Possession::
-                                       possession_buffer_time_seconds.value()));
+        Timestamp::fromSeconds(2 + Util::DynamicParameters->getEvaluationConfig()
+                                       ->getPossessionConfig()
+                                       ->PossessionBufferTimeSeconds()
+                                       ->value()));
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/evaluation/possession_test.cpp]: Like Gareth mentioned, this would be much nicer if we stored `Util::DynamicParameters->getEvaluationConfig()->getPossessionConfig()->PossessionBufferTimeSeconds()->value()));` once at the top of this function and reused that value

File: src/software/util/parameter/config_test.cpp (hunk 4 above)
Comment (jonathanlew) [reply src/software/util/parameter/config_test.cpp:241]: 💯

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Looks good! Suggested coding convention wherever we use dynamic params: we access and store all dynamic params at the top of each function, and so dynamic params are never inlined

File: src/WORKSPACE (hunk 11 above)
Comment (akhilveeraghanta) [reply src/WORKSPACE]: I just picked the most recent commit to freeze it there, these is no specific reason other than that I know the test passes w/ this library

Comment (jonathanlew) [reply src/WORKSPACE]: let's add a comment to make that clear

File: src/software/ai/hl/stp/evaluation/possession_test.cpp (hunk 12 above)
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/evaluation/possession_test.cpp]: I agree, but this is a lot of work to do across all the files, to a point to where you might as well do dependency injection at the same time if we are going through the trouble of splitting it up.

I think this file is especially bad so I can do this one, but across everything is a lot of work to do as part of this PR

Comment (jonathanlew) [reply src/software/ai/hl/stp/evaluation/possession_test.cpp]: I agree, out of scope, file an issue

File: src/WORKSPACE (hunk 11 above)
Comment (akhilveeraghanta) [reply src/WORKSPACE]: I went to write `# commit points to last known working version of library` but I don't know if that adds too much value as that is the intention of specifying the commit hash. If you think otherwise I can go back and add it as it doesn't do any harm

File: src/software/ai/hl/stp/evaluation/possession_test.cpp (hunk 12 above)
Comment (akhilveeraghanta) [reply src/software/ai/hl/stp/evaluation/possession_test.cpp]: There is this ticket for dependency injection #960 which I think will address this by nature of splitting things up

**Comment by garethellis0** [comment]
> Looks good! Suggested coding convention wherever we use dynamic params: we access and store all dynamic params at the top of each function, and so dynamic params are never inlined

This is probably a good idea for now, but it's worth noting that once we migrate to doing this via dependency injection, accessing params will look a lot more like `config->MyParam->value()`, so won't be nearly as bad.

File: src/software/util/parameter/config_test.cpp (hunk 4 above)
Comment (garethellis0) [reply src/software/util/parameter/config_test.cpp:241]: Again with that :b:. I swear I'm not :older_man: but I still don't get it :confused:

File: src/software/ai/hl/stp/evaluation/indirect_chip.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/software/ai/hl/stp/evaluation/indirect_chip.cpp]: Across all the files, just noted it here as an example.

**Comment by jonathanlew** [comment]
> > Looks good! Suggested coding convention wherever we use dynamic params: we access and store all dynamic params at the top of each function, and so dynamic params are never inlined
> 
> This is probably a good idea for now, but it's worth noting that once we migrate to doing this via dependency injection, accessing params will look a lot more like `config->MyParam->value()`, so won't be nearly as bad.

I see

File: src/software/ai/hl/stp/evaluation/possession_test.cpp (hunk 12 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/evaluation/possession_test.cpp]: Cool

**Comment by garethellis0** [comment]
(Just to be totally clear, I totally agree with the suggested convention until we make that switch)

File: src/WORKSPACE (hunk 11 above)
Comment (jonathanlew) [reply src/WORKSPACE]: Well if someone wants to change it in the future for some reason they would want to know that this is an arbitrary commit and we're not using this commit for a specific reason

Comment (garethellis0) [reply src/WORKSPACE]: That's what we're doing for all the external stuff here, should we add comments for all of them? Seems a bit overkill, it's fairly standard bazel convention to always fix your commits so the build remains hermetic.

Comment (jonathanlew) [reply src/WORKSPACE]: Ok so the default assumption is that it's some arbitrary commit

**Review by jonathanlew — DISMISSED** [review]
Address clean up in later PR?

**Comment by garethellis0** [comment]
While we're already changing things here, I think at least making the variables is in scope for this one?

**Comment by MathewMacDougall** [comment]
That might be a lot of changes? Feels a little out of scope given the size

**Comment by garethellis0** [comment]
Ok, fair enough, let's file a ticket to clean it up later. Good difficulty 1 task.

**Comment by akhilveeraghanta** [comment]
@garethellis0  filed #990

**Comment by MathewMacDougall** [comment]
LGTM pending CI

**Comment by MathewMacDougall** [comment]
Ah sorry @akhilveeraghanta looks like I just caused more conflicts. Maybe wait until #972 is merged just in case, then it should be safe to resolve. (Should be merged early Saturday morning).

**Comment by akhilveeraghanta** [comment]
Sounds good, looks like that PR only uses 1 param shouldn't be too bad

**Review by MathewMacDougall — APPROVED** [review]

**Review by jonathanlew — APPROVED** [review]

**Comment by MathewMacDougall** [comment]
@garethellis0 for approval

**Review by garethellis0 — APPROVED** [review]
